test/AcceleratingDistributor.ClaimAndStake.ts

```diff
@@ -0,0 +1,162 @@
+      {
+        windowIndex: 1,
+        account: reward2Recipients[0].account,
+        accountIndex: reward2Recipients[0].accountIndex,
+        amount: reward2Recipients[0].amount,
+        merkleProof: merkleTree2.getProof(createLeaf(reward2Recipients[0])),
+      },
+    ];
+    expect(await lpToken1.balanceOf(claimer.address)).to.equal(toBN(0));
+
+    // Tests require staker to have approved contract
+    await lpToken1.connect(claimer).approve(distributor.address, MAX_UINT_VAL);
+  });
+
+  it("Happy path", async function () {
+    const time = await distributor.getCurrentTime();
+    await expect(distributor.connect(claimer).claimAndStake(batchedClaims, lpToken1.address))
+      .to.emit(distributor, "Stake")
+      .withArgs(lpToken1.address, claimer.address, totalBatchRewards, time, totalBatchRewards, totalBatchRewards);
+    expect((await distributor.getUserStake(lpToken1.address, claimer.address)).cumulativeBalance).to.equal(
+      totalBatchRewards
+    );
+    expect(await lpToken1.balanceOf(merkleDistributor.address)).to.equal(toBN(0));
+    expect(await lpToken1.balanceOf(claimer.address)).to.equal(toBN(0));
+  });
+  it("Fails if AcceleratingDistributor is not whitelisted claimer on MerkleDistributor", async function () {
+    await merkleDistributor.whitelistClaimer(distributor.address, false);
+    await expect(distributor.connect(claimer).claimAndStake(batchedClaims, lpToken1.address)).to.be.revertedWith(
+      "invalid claimer"
+    );
+  });
+  it("MerkleDistributor set to invalid address", async function () {
+    await distributor.setMerkleDistributor(distributor.address);
+    // distributor is not a valid MerkleDistributor and error explains that.
+    await expect(distributor.connect(claimer).claimAndStake(batchedClaims, lpToken1.address)).to.be.revertedWith(
+      "function selector was not recognized and there's no fallback function"
+    );
+  });
+  it("Only owner can set MerkleDistributor address", async function () {
+    await expect(distributor.connect(claimer).setMerkleDistributor(distributor.address)).to.be.revertedWith(
+      "Ownable: caller is not the owner"
+    );
+  });
+  it("One claim account is not caller", async function () {
+    // Claiming with account that isn't receiving the claims causes revert
+    await expect(
+      distributor.connect(contractCreator).claimAndStake(batchedClaims, lpToken1.address)
+    ).to.be.revertedWith("claim account not caller");
+  });
+  it("One claim reward token is not staked token", async function () {
+    // Enable new staking token that doesn't match claims.
+    await distributor.configureStakingToken(
+      lpToken2.address,
+      true,
+      baseEmissionRate,
+      maxMultiplier,
+      secondsToMaxMultiplier
+    );
+    await expect(distributor.connect(claimer).claimAndStake(batchedClaims, lpToken2.address)).to.be.revertedWith(
+      "unexpected claim token"
+    );
+  });
+  it("Claimed token is not eligible for staking", async function () {
```

Comment:
Do we also need to add a test case for when "unexpected claim token"?

Reply:
Yup isn't that on line 146?

Reply:
I missed that somehow :) Thanks for pointing it out